sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -790,7 +790,8 @@ case class AdaptiveSparkPlanExec(
     currentPhysicalPlan.foreach {
       // earlyFailedStage is the stage which failed before calling doMaterialize,
       // so we should avoid calling cancel on it to re-trigger the failure again.
-      case s: ExchangeQueryStageExec if !earlyFailedStage.contains(s.id) =>
+      case s: ExchangeQueryStageExec if !earlyFailedStage.contains(s.id) &&
+        s.isMaterializationStarted() =>
```

Comment:
Do we still need this change if all `cancel` implementations already check this flag?

Reply:
I have added this check in terms of https://github.com/apache/spark/pull/45234#discussion_r1512850096. 
However, `cancel()` functions already has this check so we can continue with that.